### ISSUE TITLE
Type-check boolean expression coercion

### ISSUE THREAD

**Issue by rethab** [rhysd/actionlint#147]
Consider the below workflow, which has two problems. I have ran this through the [actionlint playground](https://rhysd.github.io/actionlint/#eJxtUDFywzAM2/0KxEM82Q/QXcYMfUVPtulaiSq6InW5/L6S3DodupEgCYDgYBrgwfG+eH68T9b7AgAubEllrwEKdvQ0/7bATItNXg00JvqDyhTdpo6DQXutR+CI2UktdaVDqz2uIn0lFzM7FuvlxabPjQxGZk82NM2Nx+rnIxLp7iSmIH3RSmMKmnpvlUTrSJS2w35fNg1oWhmdMULac9L8IIL9pIsdJ2OqdneIu2xH7m77h+ENsnLyc4Fe+0v+eA9t+EkLpwu6kk+H8xmn6mgonMMuLkNOsf0GnstzZQ==) but no problems were found.

```yaml
on:
  workflow_call:
    inputs:
      enabled:
        default: true
        description: "Enable or disable the workflow"
        required: false
        type: boolean

jobs:
  greet:
    runs-on: ubuntu-latest
    steps:
      - run: echo '::set-output name=abc::false'
        id: skip
      - run: echo I should run
        if: "inputs.enabled != 'true' && !steps.skip.outputs.def"
```

- Problem 1: **Inputs are actual booleans**, but the expression compares `enabled` to `'true'` which is a string. This will never be true
- Problem 2: **Outputs are always strings**, but the expression uses a the boolean not operator `!` which means `'false'` is treated as a truthy value and therefore `!('false')` is actually `false`

GitHub Docs on Expressions: https://docs.github.com/en/actions/learn-github-actions/expressions#operators

Could actionlint be extended to catch these cases as well?

**Comment by rhysd** [comment]
Yes, those type checks are not strict. But it is intended. Comparing different type values are allowed. Since actionlint is not opinionated, it should allow such loose comparison.

https://docs.github.com/en/actions/learn-github-actions/expressions#operators

> If the types do not match, GitHub coerces the type to a number. GitHub casts data types to a number using these conversions

> Outputs are always strings

There is no way to detect types of outputs statically.

**Comment by rethab** [comment]
Hi @rhysd thanks for considering this issue. I'm not sure I follow your reasoning in your response.

> Comparing different type values are allowed. Since actionlint is not opinionated [..]

Consider an expression like `inputs.enabled != 'true'`. Iff actionlint knows that the left hand side is a boolean, then it can tell the user that this expression doesn't make sense and will always be false. I don't think this is about opinion, I think this is about telling the user about something that is allowed to write, but wrong in practice.

Or can you come up with a scenario where such an expression would be desired?


> Regarding to output types, there is no way to detect types of outputs statically.

My understanding was that they're always strings, because they're passed via environment variables. So even if you set an output to `true`, when the output is read in the next step, it will be `"true"`. Is that not true?

But I guess in this case you could ague that the user uses the `!` just to check if it's an empty value.

**Comment by rmg** [comment]
In this case, since `inputs.enabled` is a boolean, the only string literals it will ever equal are `'0'` or `'1'`.

Allowing comparisons of different types definitely makes sense, but I think in the case of literals it might make sense to give a warning of some kind when the equality will always fail.

**Comment by doctau** [comment]
I'd be +1 on at least handling the specific case of comparing a boolean to a literal string (other than '0' and '1').

It is quite easy to accidentally do that when refactoring out a local reusable workflow:
1) Outputs are always strings, so you have `jobs.xxx.output.yyy == 'true'` somewhere
2) You refactor that inside a reusable workflow, and without thinking put `type: boolean` on the input
3) The value is automatically coerced from string to boolean to pass as input to the workflow, but then the comparison fails because it coerces them to numbers rather than coercing back.
